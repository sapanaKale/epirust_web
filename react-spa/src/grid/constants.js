export const BaseColors = {
    YELLOW: "#fcce44",
    LIGHT_YELLOW: "#fff0c2",
    PURPLE: "#3498db",
    BLUE: "#007bff",
    ORANGE: "#ffa500",
    RED: "#e74c3c",
    GRAY_LIGHTER: "#f1f1f1",
    PINK: "#e83e8c",
    LIGHT_PINK: "#f9d1e4",
    GRAY: "#ccc",
    BLACK: "black",
    GREEN: "#28a745",
    WHITE: "#fff",
    LIGHT_ORANGE: "#fad5b5",
    LIGHT_BLUE: "#b9dafd"
};

export const AreaColors = {
    HOUSING: BaseColors.LIGHT_YELLOW,
    WORK: BaseColors.LIGHT_PINK,
    TRANSPORT: BaseColors.LIGHT_ORANGE,
    HOSPITAL: BaseColors.LIGHT_BLUE,
    OTHER: BaseColors.GRAY
};

export const LandmarkColors = {
    HOUSES: BaseColors.GRAY,
    OFFICES: BaseColors.GRAY
};

export const AgentStateToColor = {
    's': BaseColors.BLUE,
    'e': BaseColors.ORANGE,
    'i': BaseColors.RED,
    'r': BaseColors.GREEN,
    'd': BaseColors.BLACK
};

export const AgentStateMapper = {
    's': 'susceptible',
    'e': 'exposed',
    'i': 'infected',
    'r': 'recovered',
    'd': 'deceased'
};

export const Interventions = Object.freeze({
    LOCKDOWN: 'lockdown',
    BUILD_NEW_HOSPITAL: 'build_new_hospital',
    VACCINATION: 'vaccination',
    status: {
        LOCKDOWN_START: 'locked_down',
    }
});
