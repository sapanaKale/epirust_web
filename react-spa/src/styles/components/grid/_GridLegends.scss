.legends-wrap {
    border: 2px solid powderblue;
    position: absolute;
    top: 240px;
    right: 32px;

    .legends {
        padding: 5px;
        list-style: none;

        .legend-item {
            margin-bottom: 5px;
        }

        .swatch {
            margin-right: 4px;
            display: inline-block;
        }
        .text {
            vertical-align: text-bottom;
            font-size: 12px;
        }

        .count {
            padding-right: 4px;
        }


        &.areas {
            .swatch {
                padding: 8px 15px;
            }
        }

        &.agents {
            .swatch {
                padding: 5px;
                border-radius: 50%;
            }
        }
    }
}
