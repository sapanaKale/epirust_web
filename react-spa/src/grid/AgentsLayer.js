/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import React, { useContext, useEffect, useRef, useState } from 'react';
import { GridContext } from './index'
import { AgentStateMapper, AgentStateToColor } from './constants';
import GridLegend from "./GridLegend";
import {faPauseCircle, faPlayCircle, faRedo, faStepForward, faStepBackward} from "@fortawesome/free-solid-svg-icons";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";

export default function AgentPositionsWrapper({ agentPositions, simulationEnded }) {
    const [displayIndexIncrementPaused, setDisplayIndexIncrementPaused] = useState(true);
    const [currentDisplayIndex, setCurrentDisplayIndex] = useState(0);
    const [intervalId, setIntervalId] = useState(null);
    const [clickedPause, setClickedPause] = useState(false);

    const stopIncrement = () => {
        clearInterval(intervalId);
        setIntervalId(null);
    };

    const displayedAll = () => {
        return simulationEnded && (currentDisplayIndex >= agentPositions.length - 1)
    };

    //displayed all the data
    useEffect(() => {
        if (displayedAll()) {
            stopIncrement()
        }
    }, [currentDisplayIndex, agentPositions]);

    //pause
    useEffect(() => {
        if (displayIndexIncrementPaused && agentPositions) {
            stopIncrement();
        }
    }, [displayIndexIncrementPaused, agentPositions]);

    //race condition
    useEffect(() => {
        if (!agentPositions || simulationEnded) {
            return
        }
        if (currentDisplayIndex >= agentPositions.length - 1) {
            setDisplayIndexIncrementPaused(true);
            return
        }

        if (displayIndexIncrementPaused && !clickedPause) {
            setDisplayIndexIncrementPaused(false)
        }

    }, [currentDisplayIndex, agentPositions, simulationEnded, clickedPause]);

    //count++
    useEffect(() => {
        if (displayIndexIncrementPaused)
            return;

        if (!agentPositions || intervalId)
            return;

        const interval = setInterval(() => setCurrentDisplayIndex(index => index + 1), 100);
        console.log("interval started", interval);
        setIntervalId(interval);

        return () => clearInterval(intervalId)
    }, [displayIndexIncrementPaused, agentPositions]);

    const handleResume = () => {
        setClickedPause(false);
        setDisplayIndexIncrementPaused(false)
    };

    const handlePause = () => {
        setClickedPause(true);
        setDisplayIndexIncrementPaused(true)
    };

    const handleReset = () => {
        setCurrentDisplayIndex(0);
        setClickedPause(true);
        setDisplayIndexIncrementPaused(true)
    };

    const positionsToDisplay = agentPositions
        ? agentPositions[currentDisplayIndex]
        : [];

    const getCounts = (citizens) => {
        const counts = {
            'susceptible': 0,
            'exposed': 0,
            'infected': 0,
            'recovered': 0,
            'deceased': 0
        };
        citizens && citizens.forEach(a => counts[AgentStateMapper[a.state]] += 1);
        return counts

    };

    const handleAddHour = () => {
        if (currentDisplayIndex + 1 >= agentPositions.length)
            return

        setCurrentDisplayIndex(idx => idx + 1)
    }

    const handleSubtractHour = () => {
        if (currentDisplayIndex <= 0)
            return

        setCurrentDisplayIndex(idx => idx - 1)
    }

    const dayAndHour = (index) => {
        let day = Math.floor((index + 1) / 24);
        let hour = (index + 1) % 24;
        return `Day ${day}, hour: ${hour}`;
    }

    return (
        <div style={{ position: "relative" }}>
            <div className="card p-4 grid-control" style={{ position: "absolute", zIndex: 5, right: 0, width: 190 }}>
                <h4 style={{ textAlign: 'center' }} className="pb-2" data-testid="counter">{`${currentDisplayIndex + 1}/${agentPositions ? agentPositions.length : 0} hrs`}</h4>

                <div style={{ textAlign: 'center' }} className="pb-2" >{ dayAndHour(currentDisplayIndex) }</div>

                <div className="btn-group mb-2 justify-content-center" role="group" aria-label="Buttons to pause/play and reset">
                    {clickedPause
                        ? <FontAwesomeIcon onClick={handleResume} data-testid="resume" disabled={displayedAll()} icon={faPlayCircle} className="icon resume"/>
                        : <FontAwesomeIcon data-testid="pause" icon={faPauseCircle} onClick={handlePause} disabled={displayedAll()} className="icon pause"/>
                    }

                    <FontAwesomeIcon onClick={handleReset} data-testid="reset" icon={faRedo} className="icon reset"/>
                </div>

                {(clickedPause || displayedAll()) && (
                    <div data-testid="hour-step-controls" className="btn-group my-0 mx-auto hour-step-control" role="group" aria-label="Buttons to increase/decrease current displayed hour">
                        <FontAwesomeIcon icon={faStepBackward} data-testid="decrement-hour" name="decrement hour" onClick={handleSubtractHour} className="icon"/>
                        <FontAwesomeIcon icon={faStepForward} data-testid="increment-hour" name="increment hour" onClick={handleAddHour} className="icon" />
                    </div>
                )}

            </div>
            <AgentsLayer agentPositionsPerHour={positionsToDisplay} />

            <GridLegend {...getCounts(positionsToDisplay)} />
        </div>
    )
}

function AgentsLayer({ agentPositionsPerHour }) {
    const { cellDimension, lineWidth, canvasDimension } = useContext(GridContext);

    const agentsLayerCanvas = useRef(null);
    const [agentsCanvasContext, setAgentsCanvasContext] = useState(null);

    useEffect(() => {
        if (!agentsLayerCanvas)
            return;

        setAgentsCanvasContext(agentsLayerCanvas.current.getContext("2d"));

    }, [agentsLayerCanvas]);

    useEffect(() => {
        if (!agentsCanvasContext || !agentPositionsPerHour)
            return;

        agentsCanvasContext.clearRect(0, 0, canvasDimension, canvasDimension);

        agentPositionsPerHour.forEach((agent) => {
            const { x, y } = agent.location;

            agentsCanvasContext.fillStyle = AgentStateToColor[agent.state];

            agentsCanvasContext.beginPath();
            const agentCircleRadius = Math.floor(cellDimension / 2);
            const startAngle = 0, endAngle = 2 * Math.PI;
            agentsCanvasContext.arc(calculateCoordinate(x), calculateCoordinate(y), agentCircleRadius, startAngle, endAngle);
            agentsCanvasContext.fill();
        })
    }, [agentsCanvasContext, agentPositionsPerHour]);

    function calculateCoordinate(x) {
        return Math.floor(x * cellDimension + (0.5 * cellDimension) + lineWidth)
    }

    return (
        <canvas ref={agentsLayerCanvas} data-testid="grid-canvas-agents" id="grid-canvas-agents" width={canvasDimension}
            height={canvasDimension} style={{ position: "absolute", zIndex: 4 }} />
    )
}
