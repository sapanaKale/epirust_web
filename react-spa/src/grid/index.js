/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from 'react';

import LandmarksLayer from './LandmarksLayer';
import LinesLayer from './LinesLayer';
import AreasLayer from './AreasLayer';
import AgentsLayer from './AgentsLayer';
import {AreaColors} from './constants';
import io from 'socket.io-client'
import config from "../config";
import Loader from '../common/Loader';
import PropTypes from "prop-types";

export const GridContext = React.createContext(null);

export default function GridPage({jobId}) {
    const [socket, setSocket] = useState(null);
    const [socketDataExhausted, setSocketDataExhausted] = useState(false);

    //default values?
    const [areaDimensions, setAreaDimensions] = useState(null);
    const [landmarksDimensions, setLandmarksDimensions] = useState(null);
    const [agentPositions, setAgentPositions] = useState(null);

    const [gridContextData, setGridContextData] = useState(null);

    useEffect(() => {
        console.log("started socket");
        const socketInstance = io(`${config.API_HOST}/grid-updates`);
        setSocket(socketInstance);

        return () => {
            socketInstance.close();
        }
    }, []);

    //reading socket data
    useEffect(() => {
        if (!socket)
            return;

        socket.emit('get', {jobId});

        socket.on('gridData', function (message) {
            if ("simulation_ended" in message) {
                socket.close();
                setSocketDataExhausted(true);
                return
            }

            if ('grid_size' in message) {
                const { housing_area, work_area, transport_area, hospital_area, grid_size } = message;
                const areaDimensions = [
                    { ...housing_area, color: AreaColors.HOUSING },
                    { ...work_area, color: AreaColors.WORK },
                    { ...transport_area, color: AreaColors.TRANSPORT },
                    { ...hospital_area, color: AreaColors.HOSPITAL }
                ];

                const housesDimensions = message.houses,
                    officesDimensions = message.offices;

                const cellDimension = Math.floor((window.innerHeight - 150) / grid_size),
                    lineWidth = Math.floor(cellDimension / 4) < 1 ? 0 : Math.floor(cellDimension / 4),
                    canvasDimension = (grid_size * cellDimension) + lineWidth;

                setGridContextData({
                    cellDimension: cellDimension,
                    lineWidth: lineWidth,
                    canvasDimension: canvasDimension,
                    size: grid_size
                });

                setAreaDimensions(areaDimensions);
                setLandmarksDimensions({ housesDimensions, officesDimensions });
                return
            }

            if ('citizen_states' in message) {

                setAgentPositions(pos => {
                    if (!pos)
                        return [message.citizen_states];

                    return [...pos, message.citizen_states]
                })
            }

        });
    }, [socket, jobId]);

    if (!gridContextData)
        return (
          <div className="grid-wrap">
              <Loader />
          </div>
        )

    return (
        <div className="grid-wrap">
            <GridContext.Provider value={gridContextData}>
                <div style={{ position: "relative" }}>
                    <AreasLayer areaDimensions={areaDimensions} />
                    <LinesLayer />
                    <LandmarksLayer landmarksDimensions={landmarksDimensions} />
                    {<AgentsLayer agentPositions={agentPositions} simulationEnded={socketDataExhausted} />}
                </div >
            </GridContext.Provider>
        </div>
    )
}

GridPage.propTypes = {
    jobId: PropTypes.string.isRequired
}