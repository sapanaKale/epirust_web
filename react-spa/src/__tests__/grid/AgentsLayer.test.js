/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from 'react'
import { act, fireEvent, render } from '@testing-library/react'
import { GridContext } from '../../grid/index'

import AgentPositionsWrapper from '../../grid/AgentsLayer';
import { element } from 'prop-types';

jest.useFakeTimers();

afterEach(() => {
    jest.clearAllTimers();
    jest.clearAllMocks();
});

const gridContextData = {
    cellDimension: 2,
    lineWidth: 1,
    canvasDimension: 20,
    size: 10
};
const agentPositions = [[
    { "id": 595, "state": "s", "location": { "x": 0, "y": 0 } },
    { "id": 238, "state": "s", "location": { "x": 37, "y": 66 } },
    { "id": 981, "state": "s", "location": { "x": 31, "y": 1 } }
]];

test('should pause without clicking when displayed all hours are sent by socket and resume when new data is received', () => {
    const { getByTestId, rerender } = render(
        <GridContext.Provider value={gridContextData}>
            <AgentPositionsWrapper agentPositions={[agentPositions[0], agentPositions[0]]} simulationEnded={false} />
        </GridContext.Provider>);

    act(() => {
        jest.advanceTimersByTime(500)
    });

    expect(getByTestId("counter").textContent).toBe("2/2 hrs");
    expect(clearInterval).toHaveBeenCalled();

    rerender(
        <GridContext.Provider value={gridContextData}>
            <AgentPositionsWrapper agentPositions={[agentPositions[0], agentPositions[0], agentPositions[0]]} simulationEnded={false} />
        </GridContext.Provider>);

    expect(setInterval).toHaveBeenCalled();

    act(() => {
        jest.advanceTimersByTime(100)
    });

    expect(getByTestId("counter").textContent).toBe("3/3 hrs");
    expect(getByTestId("grid-canvas-agents").getContext("2d").__getEvents()).toMatchSnapshot()
});

test('should render AgentsLayer when grid visualization resets', () => {

    const { getByTestId} = render(
        <GridContext.Provider value={gridContextData}>
            <AgentPositionsWrapper agentPositions={agentPositions} simulationEnded={false} />
        </GridContext.Provider>);
    act(() => {
        jest.advanceTimersByTime(100)
    });

    expect(getByTestId("counter").textContent).toBe("1/1 hrs");

    act(() => {
        fireEvent.click(getByTestId("reset"));
    });

    expect(getByTestId("resume")).toBeInTheDocument();
    expect(getByTestId("counter").textContent).toBe("1/1 hrs");
    expect(getByTestId("grid-canvas-agents").getContext("2d").__getEvents()).toMatchSnapshot()
});

test('should render AgentsLayer when grid visualization pauses and resume', () => {

    const { getByTestId, rerender } = render(
        <GridContext.Provider value={gridContextData}>
            <AgentPositionsWrapper agentPositions={[agentPositions[0], agentPositions[0], agentPositions[0]]} simulationEnded={false} />
        </GridContext.Provider>);

    act(() => {
        jest.advanceTimersByTime(100);
        fireEvent.click(getByTestId("pause"))
    });

    expect(getByTestId("counter").textContent).toBe("2/3 hrs");

    act(() => {
        fireEvent.click(getByTestId("resume"))
    });

    rerender(
        <GridContext.Provider value={gridContextData}>
            <AgentPositionsWrapper agentPositions={[agentPositions[0], agentPositions[0], agentPositions[0]]} simulationEnded={true} />
        </GridContext.Provider>
    );

    act(() => {
        jest.advanceTimersByTime(100)
    });

    expect(getByTestId("counter").textContent).toBe("3/3 hrs");
    expect(getByTestId("grid-canvas-agents").getContext("2d").__getEvents()).toMatchSnapshot()
});

test('should display hour step controls i.e. buttons to increment/decrement on pausing the simulation', () => {

    const { getByTestId, queryByTestId} = render(
        <GridContext.Provider value={gridContextData}>
            <AgentPositionsWrapper agentPositions={[agentPositions[0], agentPositions[0], agentPositions[0]]} simulationEnded={false} />
        </GridContext.Provider>);

    expect(queryByTestId("hour-step-controls")).not.toBeInTheDocument()

    act(() => {
        jest.advanceTimersByTime(100);
        fireEvent.click(getByTestId("pause"))
    });

    expect(getByTestId("hour-step-controls")).toBeInTheDocument()

    act(() => {
        fireEvent.click(getByTestId("resume"))
    });

    expect(queryByTestId("hour-step-controls")).not.toBeInTheDocument()
});

test('should display hour step controls i.e. buttons to increment/decrement when displayed all hours', () => {

    const { getByTestId, queryByTestId, rerender } = render(
        <GridContext.Provider value={gridContextData}>
            <AgentPositionsWrapper agentPositions={[agentPositions[0], agentPositions[0], agentPositions[0]]} simulationEnded={false} />
        </GridContext.Provider>);

    act(() => {
        jest.advanceTimersByTime(100);
    });

    expect(queryByTestId("hour-step-controls")).not.toBeInTheDocument();
    expect(getByTestId("counter").textContent).toBe("2/3 hrs");

    act(() => {
        jest.advanceTimersByTime(100);
    });

    rerender(
        <GridContext.Provider value={gridContextData}>
            <AgentPositionsWrapper agentPositions={[agentPositions[0], agentPositions[0], agentPositions[0]]} simulationEnded={true} />
        </GridContext.Provider>
    );

    expect(getByTestId("hour-step-controls")).toBeInTheDocument()
    expect(getByTestId("counter").textContent).toBe("3/3 hrs");
});

test('should increment currently displayed hour value on clicking + button', () => {

    const { getByTestId} = render(
        <GridContext.Provider value={gridContextData}>
            <AgentPositionsWrapper agentPositions={[agentPositions[0], agentPositions[0], agentPositions[0]]} simulationEnded={false} />
        </GridContext.Provider>);

    act(() => {
        jest.advanceTimersByTime(100);
        fireEvent.click(getByTestId("pause"))
    });
    expect(getByTestId("counter").textContent).toBe("2/3 hrs");

    act(() => { fireEvent.click(getByTestId("decrement-hour")) });
    expect(getByTestId("counter").textContent).toBe("1/3 hrs");
    const canvasEventsHour1 = getByTestId("grid-canvas-agents").getContext("2d").__getEvents()

    act(() => { fireEvent.click(getByTestId("increment-hour")) });
    expect(getByTestId("counter").textContent).toBe("2/3 hrs");

    act(() => { fireEvent.click(getByTestId("decrement-hour")) });
    expect(getByTestId("counter").textContent).toBe("1/3 hrs");
    const canvasEventsHour1Rerendered = getByTestId("grid-canvas-agents").getContext("2d").__getEvents()

    expect(canvasEventsHour1).toEqual(canvasEventsHour1Rerendered)
});
