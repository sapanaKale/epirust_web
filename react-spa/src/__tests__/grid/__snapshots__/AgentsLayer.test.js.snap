// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pause without clicking when displayed all hours are sent by socket and resume when new data is received 1`] = `
Array [
  Object {
    "props": Object {
      "height": 20,
      "width": 20,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "value": "#007bff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 1,
      "startAngle": 0,
      "x": 2,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 1,
            "startAngle": 0,
            "x": 2,
            "y": 2,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "value": "#007bff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 1,
      "startAngle": 0,
      "x": 76,
      "y": 134,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 1,
            "startAngle": 0,
            "x": 76,
            "y": 134,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "value": "#007bff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 1,
      "startAngle": 0,
      "x": 64,
      "y": 4,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 1,
            "startAngle": 0,
            "x": 64,
            "y": 4,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;

exports[`should render AgentsLayer when grid visualization pauses and resume 1`] = `
Array [
  Object {
    "props": Object {
      "height": 20,
      "width": 20,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "value": "#007bff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 1,
      "startAngle": 0,
      "x": 2,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 1,
            "startAngle": 0,
            "x": 2,
            "y": 2,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "value": "#007bff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 1,
      "startAngle": 0,
      "x": 76,
      "y": 134,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 1,
            "startAngle": 0,
            "x": 76,
            "y": 134,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "value": "#007bff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 1,
      "startAngle": 0,
      "x": 64,
      "y": 4,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 1,
            "startAngle": 0,
            "x": 64,
            "y": 4,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;

exports[`should render AgentsLayer when grid visualization resets 1`] = `
Array [
  Object {
    "props": Object {
      "height": 20,
      "width": 20,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "value": "#007bff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 1,
      "startAngle": 0,
      "x": 2,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 1,
            "startAngle": 0,
            "x": 2,
            "y": 2,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "value": "#007bff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 1,
      "startAngle": 0,
      "x": 76,
      "y": 134,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 1,
            "startAngle": 0,
            "x": 76,
            "y": 134,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "value": "#007bff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 1,
      "startAngle": 0,
      "x": 64,
      "y": 4,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 1,
            "startAngle": 0,
            "x": 64,
            "y": 4,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;
